### PR TITLE
Download() refactor

### LINKED ISSUE #179: Refactor `dataset.download`
* define a global downloads variable which is a dictionary mapping from download type to a dictionary of key value pairs of identifier: download object
* Allow partial download of objects by specifying a list of key - by default download everything (possibly simplify the whole `download` function and `download_utils.downloader` function)
* Add a mandatory print statement controlled by an argument `verbose=True` which is printed before downloading before downloading begins with information about what is happening.

### PR DESCRIPTION
**Pull request by magdalenafuentes**
### Description

Refactor of the `downloader()` function to allow partial downloads, and improve its transparency.

Fixes #179, #168, and #244 

This PR does:

- [x] Add global variable `REMOTES`, which is a dictionary that contains the different `RemoteFileMetadata` files.
- [x] Simplify download functions in modules
- [x] Adds message when download starts explaining what's being downloaded and where
- [x] Prints `url` when `mirdata` fails to download
- [x] Update tests_download_utils.py
- [x] Update `CONTRIBUTING.md`
- [x] Customize download functions to make sense for each loader

### CODE REVIEW

**Comment by magdalenafuentes** [comment]
Still have to update the tests, but will wait for @rabitt and @lostanlen feedback before

File: mirdata/download_utils.py [hunk 1]
```diff
@@ -60,6 +48,38 @@ def downloader(
...
         os.makedirs(save_dir)
 
+    if type(download_items) != dict:
+        raise TypeError(
+            'download_items should be a dictionary, but is '
+            + 'a {} element'.format(type(download_items))
+        )
+
+    start_download_message = """
+            Trying to download the list of files {} to folder {}
```
Comment (rabitt) [inline mirdata/download_utils.py:58]: nit: Trying --> Starting

File: mirdata/gtzan_genre.py [hunk 2]
```diff
@@ -152,8 +153,12 @@ def download(data_home=None):
     """
     if data_home is None:
         data_home = utils.get_default_dataset_path(DATASET_DIR)
+    if download_items is None:
```
Comment (rabitt) [inline mirdata/gtzan_genre.py:156]: with my edit above, this code can go away!

File: mirdata/beatles.py [hunk 3]
```diff
@@ -17,12 +17,14 @@
 import mirdata.jams_utils as jams_utils
 
 DATASET_DIR = 'Beatles'
-ANNOTATIONS_REMOTE = download_utils.RemoteFileMetadata(
-    filename='The Beatles Annotations.tar.gz',
-    url='http://isophonics.net/files/annotations/The%20Beatles%20Annotations.tar.gz',
-    checksum='62425c552d37c6bb655a78e4603828cc',
-    destination_dir='annotations',
-)
+DOWNLOAD = {
```
[the remark this answers is not in the document]
Comment (lostanlen) [reply mirdata/beatles.py:20]: REMOTES ?

File: mirdata/beatles.py [hunk 4]
```diff
@@ -17,12 +17,14 @@
...
+    'ANNOTATIONS_REMOTE': download_utils.RemoteFileMetadata(
```
[the remark this answers is not in the document]
Comment (lostanlen) [reply mirdata/beatles.py:21]: it's also the name of the subdirectory :)

File: mirdata/beatles.py [hunk 5]
```diff
@@ -121,20 +123,25 @@ def load_audio(audio_path):
...
-def download(data_home=None, force_overwrite=False):
+def download(data_home=None, download_items=None, force_overwrite=False):
     """Download the Beatles Dataset (annotations).
     The audio files are not provided due to the copyright.
 
     Args:
         data_home (str): Local path where the dataset is stored.
             If `None`, looks for the data in the default directory, `~/mir_datasets`
         force_overwrite (bool): Whether to overwrite the existing downloaded data
+        to_download (dict): Remote objects to download. By default it will download all available objects of a given
```
[the remark this answers is not in the document]
Comment (magdalenafuentes) [reply mirdata/beatles.py:134]: I would go for uniformity maybe, and add it to all

[the remark this answers is not in the document]
Comment (rabitt) [reply mirdata/beatles.py:134]: We're not totally uniform right now (e.g. `force_overwrite` isn't part of every `download` function). If you want to go that way, maybe we should change it for everything?

File: mirdata/download_utils.py [hunk 6]
```diff
@@ -24,9 +24,8 @@
 
 def downloader(
     save_dir,
-    zip_downloads=None,
-    tar_downloads=None,
-    file_downloads=None,
+    download=None,
```
Comment (rabitt) [inline mirdata/download_utils.py:27]: maybe we can call it `remotes` to match the naming in the modules?

File: mirdata/download_utils.py [hunk 7]
```diff
@@ -57,20 +53,59 @@ def downloader(
...
+        if partial_download is not None:
+            if type(partial_download) != list:
+                raise TypeError(
+                    'partial_download should be a list of keys to access the elements of the download dictionary, '
+                    + 'but is a {} element'.format(type(partial_download))
+                )
+            # check the keys in partial_download are in the download dict
+            for key in partial_download:
+                if key not in download.keys():
+                    raise KeyError(
```
Comment (rabitt) [inline mirdata/download_utils.py:66]: '{} not found. List items in `partial_download` should be one of {}`.format(key, download.keys()`

File: mirdata/download_utils.py [hunk 8]
```diff
@@ -57,20 +53,59 @@ def downloader(
...
-        for file_download in file_downloads:
-            download_from_remote(file_download, save_dir, force_overwrite)
-
-    if info_message is not None:
-        print(info_message)
+    if download is not None:
+        if partial_download is not None:
+            if type(partial_download) != list:
+                raise TypeError(
+                    'partial_download should be a list of keys to access the elements of the download dictionary, '
```
Comment (rabitt) [inline mirdata/download_utils.py:60]: `'partial_download must be a list which is a subset of {}'.format(download.keys()`

File: mirdata/download_utils.py [hunk 9]
```diff
@@ -57,20 +53,59 @@ def downloader(
...
-
-    if file_downloads is not None:
-        for file_download in file_downloads:
-            download_from_remote(file_download, save_dir, force_overwrite)
-
-    if info_message is not None:
-        print(info_message)
+    if download is not None:
+        if partial_download is not None:
+            if type(partial_download) != list:
```
Comment (rabitt) [inline mirdata/download_utils.py:58]: nit: `if not isinstance(partial_download, list)`

File: mirdata/download_utils.py [hunk 10]
```diff
@@ -57,20 +53,59 @@ def downloader(
...
+            for zip_download in zip_downloads:
+                download_zip_file(zip_download, save_dir, force_overwrite, cleanup)
+
+        if tar_downloads is not None:
+            for tar_download in tar_downloads:
+                download_tar_file(tar_download, save_dir, force_overwrite, cleanup)
+
+        if file_downloads is not None:
+            for file_download in file_downloads:
+                download_from_remote(file_download, save_dir, force_overwrite)
```
Comment (rabitt) [inline mirdata/download_utils.py:105]: I think this blcok can be simplified!

```python
for k in objs_to_download:
    extension = os.path.splitext(download[k].url)[-1]
    if '.zip' in extension:
        download_zip_file(download[k], save_dir, force_overwrite, cleanup)
    elif '.tar.gz' in extension or '.tar' in extension:
        download_tar_file(download[k], save_dir, force_overwrite, cleanup)
    else:
        download_from_remote(download[k], save_dir, force_overwrite, cleanup)
```

File: mirdata/download_utils.py [hunk 11]
```diff
@@ -124,12 +159,14 @@ def download_from_remote(remote, save_dir, force_overwrite=False):
...
                 error_msg = """
-                            mirdata failed to download the dataset!
+                            mirdata failed to download the dataset from {}!
                             Please try again in a few minutes.
                             If this error persists, please raise an issue at
                             https://github.com/mir-dataset-loaders/mirdata,
                             and tag it with 'broken-link'.
-                            """
+                            """.format(
+                    remote.url
```
Comment (rabitt) [inline mirdata/download_utils.py:168]: <3

File: mirdata/ikala.py [hunk 12]
```diff
@@ -25,12 +25,14 @@
...
 DATASET_DIR = 'iKala'
 TIME_STEP = 0.032  # seconds
-ID_MAPPING_REMOTE = download_utils.RemoteFileMetadata(
-    filename='id_mapping.txt',
-    url='http://mac.citi.sinica.edu.tw/ikala/id_mapping.txt',
-    checksum='81097b587804ce93e56c7a331ba06abc',
-    destination_dir=None,
-)
+REMOTES = {
+    'annotations': download_utils.RemoteFileMetadata(
```
Comment (rabitt) [inline mirdata/ikala.py:29]: 'annotations' --> 'metadata'

File: mirdata/rwc_classical.py [hunk 13]
```diff
@@ -201,13 +204,32 @@ def load_audio(audio_path):
...
+            Whether to delete the zip/tar file after extracting.
+
+    """
 
     if data_home is None:
         data_home = utils.get_default_dataset_path(DATASET_DIR)
 
     info_message = """
-        Unfortunately the audio files of the RWC-Jazz dataset are not available
+        Unfortunately the audio files of the RWC-Classical dataset are not available
```
Comment (rabitt) [inline mirdata/rwc_classical.py:232]: 👍

**Comment by magdalenafuentes** [comment]
@rabitt do you recall why did we choose to set `cleanup=False` as default?

**Comment by rabitt** [comment]
> @rabitt do you recall why did we choose to set `cleanup=False` as default?

It was so that if you downloaded e.g. multiple big zip/tar files and the download broke at some point, when you restarted it wouldn't have to start again from the beginning. But with this change maybe it makes sense to change the default?

File: mirdata/dali.py [hunk 14]
```diff
@@ -202,7 +204,7 @@ def load_audio(audio_path):
     return librosa.load(audio_path, sr=None, mono=True)
 
 
-def download(data_home=None, force_overwrite=False):
+def download(data_home=None):
```
Comment (rabitt) [inline mirdata/dali.py:207]: shall we keep force_overwrite here? since something is being downloaded

File: mirdata/groove_midi.py [hunk 15]
```diff
@@ -372,18 +374,28 @@ def load_midi(midi_path):
     return pretty_midi.PrettyMIDI(midi_path)
 
 
-def download(data_home=None):
+def download(data_home=None, force_overwrite=False, cleanup=True):
```
Comment (rabitt) [inline mirdata/groove_midi.py:377]: good catch!

File: mirdata/ikala.py [hunk 16]
```diff
@@ -202,16 +204,15 @@ def load_mix_audio(audio_path):
     return 2.0 * mixed_audio, sr
 
 
-def download(data_home=None, force_overwrite=False):
+def download(data_home=None):
```
Comment (rabitt) [inline mirdata/ikala.py:207]: force_overwrite

File: mirdata/medleydb_melody.py [hunk 17]
```diff
@@ -183,7 +184,9 @@ def download(data_home=None):
         data_home=data_home
     )
 
-    download_utils.downloader(data_home, info_message=info_message)
+    download_utils.downloader(
+        data_home, remotes=None, info_message=info_message,
```
Comment (rabitt) [inline mirdata/medleydb_melody.py:188]: nit: `remotes` isn't needed here since it defaults to None.

File: mirdata/medleydb_pitch.py [hunk 18]
```diff
@@ -156,7 +157,7 @@ def download(data_home=None, force_overwrite=False):
     )
 
     download_utils.downloader(
-        data_home, info_message=info_message, force_overwrite=force_overwrite
+        data_home, remotes=None, info_message=info_message,
```
Comment (rabitt) [inline mirdata/medleydb_pitch.py:160]: same as above

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -260,40 +262,49 @@ def load_audio(audio_path):
     return librosa.load(audio_path, sr=None, mono=True)
 
 
-def download(data_home=None, force_overwrite=False):
+def download(
```
Comment (rabitt) [inline CONTRIBUTING.md:265]: maybe we can add the following comment just above this function:

```
# -- the partial_download argument can be removed if `dataset.REMOTE` is missing/has only one value
# -- the force_overwrite argument can be removed if the dataset does not download anything (i.e. there is no `dataset.REMOTE`)
# -- the cleanup argument can be removed if the dataset has no tar or zip files in `dataset.REMOTE`.
```

**Comment by rabitt** [comment]
Thanks @magdalenafuentes for the giant PR!

**Pull request merged**